### ISSUE TITLE
Comments and line continuation don't work with "readInputLine"

### ISSUE THREAD

**Issue by raimis** [plumed/plumed2#571]
I'm working on improving the _PLUMED_ plugin for _OpenMM_ (https://github.com/openmm/openmm-plumed).

_OpenMM-PLUMED_ uses `plumed_cmd(plumedmain, "readInputLine", &line)` to pass the input file to _PLUMED_ (line by line), but the comments and line continuation don't work.

For a simple example, it works as expected:
```
d1: DISTANCE ATOMS=1,2
PRINT ARG=d1 FILE=colvar STRIDE=10 
```
If the input contains a comment, _PLUMED_ fails:
```
# comment
d1: DISTANCE ATOMS=1,2
PRINT ARG=d1 FILE=colvar STRIDE=10 
```
```
+++ PLUMED error
+++ at PlumedMain.cpp:670, function void PLMD::PlumedMain::readInputWords(const std::vector<std::__cxx11::basic_string<char> >&)
+++ message follows +++
ERROR
I cannot understand line: # comment
Maybe a missing space or a typo?
```
If the input contains a line continuation, _PLUMED_ fails also:
```
d1: DISTANCE ...
  ATOMS=1,2
... 
PRINT ARG=d1 FILE=colvar STRIDE=10 
```
```
+++ PLUMED error
+++ at Action.cpp:242, function void PLMD::Action::error(const string&) const
+++ message follows +++
ERROR in input to action DISTANCE with label d1 : Number of specified atoms should be 2
```

Would it be possible to fix `plumed_cmd(plumedmain, "readInputLine", &line)` or find another way to pass the input as a string?

P.S. Due the architecture of _OpenMM_,  `plumed_cmd(plumedmain, "setPlumedDat", &plumedInput)` isn't an option.

**Comment by GiovanniBussi** [comment]
There's no way with the current version of the code, but it can be
implemented.

Before making the effort I want to be sure this is really necessary. Can
you just dump all your lines to a temporary file and pass that to plumed?
Or do you need to read some line, run a few steps, and read some other line?

**Comment by raimis** [comment]
@GiovanniBussi 

*OpenMM-PLUMED* reads the entire file at once. You proposed solution of dumping a temporary file might work, but it is not an ideal solution: using I/O to transfer data that is already in memory.

Before starting hack around this issue, I was wondering if this is a bug or a "feature". I assume, passing a string directly instead of reading from it from a file, should be a trivial thing.

**Comment by GiovanniBussi** [comment]
If OpenMM-PLUMED reads the entire file, does this mean that the file exists somewhere? So why not passing its path instead of reading it in memory?

Cmd `readInputLine` was initially exposed for machine generated inputs (that do not require comments or continuation lines) or for MD codes that want to embed the PLUMED input within their own input, and thus have their own means to escape newlines or allow comments.

The reason why I would say this is a feature request rather than a bug fix is that this method was called `readInputLine` to allow reading one line. If we implement the possibility to have continuation lines, I would call it `readInputLines` and not change the behavior of the original method.

The reason why it requires some work is that currently the parser works like this:
- Read a line from a file, resolving comments and continuation lines.
- Call `readInputLine` for that line.

It's not difficult to implement, but if its implementation takes more time than passing the path to an existing file this would go to low priority.

**Comment by raimis** [comment]
> If OpenMM-PLUMED reads the entire file, does this mean that the file exists somewhere? So why not passing its path instead of reading it in memory?
> It's not difficult to implement, but if its implementation takes more time than passing the path to an existing file this would go to low priority.

1. *OpenMM* is a library. The file might exist or not, it depends on the calling code.
2. *OpenMM* objects are serializable, i.e. the simulation setup and state can be save to a single XML file and moved to different machine or stored. An additional external file will brake that feature.

**Comment by GiovanniBussi** [comment]
OK I see.

What I would do then is I would implement a new cmd `readInputLines` that:
1. writes the string on a temporary file
2. calls cmd("read","/path/to/that/file")

If order to use it you will have to do something like:
```
int pversion;
plumed.cmd("getApiVersion",&pversion);
if(pversion>7) { // this means only PLUMED 2.7 or later
  plumed.cmd("readInputLines,buffer);
} else {
  // the old way here
}
```

Would this work for you?

**Comment by tonigi** [comment]
I guess this will also solve the issue with `UNITS`  https://github.com/openmm/openmm-plumed/issues/32    and `ENDPLUMED` (which is ignored).

**Comment by raimis** [comment]
@GiovanniBussi thank you for implementing `readInputLines`.

Do you have any time line when these changes will be released?

**Comment by GiovanniBussi** [comment]
This will go in PLUMED 2.7, due at the end of year 2020, but will be
available in the PLUMED master branch as soon as I merge the branch (just
need some more time to set up a few more regression tests, perhaps I will
manage today).
